**Sympy** is a **Python** package used for solving equations using symbolic math. 

Let's solve the following problem with **SymPy**. 

#### Given:
The density of two different polymer samples $\rho_1$ and $\rho_2$ are measured. 

$$ \rho_1 = 0.904 \ g/cm^3 $$

$$ \rho_2 = 0.895 \ g/cm^3 $$

The percent crystalinity of the two samples ($\%c_1 $ and $\%c_2$) is known. 

$$ \%c_1 = 62.8 \% $$

$$ \%c_2 = 54.4 \% $$

The percent crystalinity of a polymer sample is related to the density of 100% amorphus regions ($\rho_a$) and 100% crystaline regions ($\rho_c$) according to:

$$ \%crystallinity = \frac{ \rho_c(\rho_s - \rho_a)   }{\rho_s(\rho_c - \rho_a)   } \times 100 \% $$

#### Find:

Find the density of 100% amorphus regions ($\rho_a$) and the density of 100% crystaline regions ($\rho_c$) for this polymer.

#### Solution:

There are a couple functions we need from **Sympy**. We'll need the ```symbols``` function to create our symbolic math variables and we need the ```nonlinsolve``` function to solve a system of non-linear equations.

In [1]:
from sympy import symbols, nonlinsolve, solve

We need to define six different symbols: $$\rho_c, \rho_a, \rho_1, \rho_2, c_1, c_2$$

In [2]:
pc, pa, p1, p2, c1, c2 = symbols('pc pa p1 p2 c1 c2')

Next we'll create two expressions for our two equations. We can subtract the %crystallinity from the left side of the equation to set the equation to zero. 

$$ \%crystallinity = \frac{ \rho_c(\rho_s - \rho_a)   }{\rho_s(\rho_c - \rho_a)   } \times 100 \% $$

$$ \frac{ \rho_c(\rho_s - \rho_a)   }{\rho_s(\rho_c - \rho_a)   } \times 100 \% - \%crystallinity = 0 $$



Sub in $\rho_s = \rho_1$ and $\rho_s = \rho_2$ to each of the expressions.

In [3]:
expr1 = ( (pc*(p1-pa)   ) / (p1*(pc-pa)) - c1)
expr2 = ( (pc*(p2-pa)   ) / (p2*(pc-pa)) - c2)

Now we'll substitue in the values of $\rho_1 = 0.904$ and $c_1 = 0.628$ into our first expression.

In [4]:
expr1 = expr1.subs(p1, 1.408)

In [5]:
expr1 = expr1.subs(c1, 0.743)

In [6]:
expr1

0.710227272727273*pc*(-pa + 1.408)/(-pa + pc) - 0.743

Now we'll substitue our the values of $\rho_2 = 0.895$ and $c_2 = 0.544$ into our second expression.

In [7]:
expr2 = expr2.subs(p2, 1.343)

In [8]:
expr2 = expr2.subs(c2, 0.312)

In [9]:
expr2

0.744601638123604*pc*(-pa + 1.343)/(-pa + pc) - 0.312

To solve the two equations for the to unknows $\rho_a$ and $\rho_b$, use SymPy's ```nonlinsolve()``` function. Pass in a list of the two expressions and followed by a list of the two variables to solve for.

In [10]:
nonlinsolve([expr1,expr2],[pa,pc])

{(1.29957020576342, 1.44984214271483)}

We see that the value of $\rho_a = 0.840789$ and $\rho_c = 0.94613431$.

The solution is a SymPy ```FiniteSet``` object. To pull the values of $\rho_a$ and $\rho_c$ out of the ```FiniteSet```, use the syntax ```sol.args[0][<var num>]```.

In [11]:
sol = nonlinsolve([expr1,expr2],[pa,pc])

In [12]:
type(sol)

sympy.sets.sets.FiniteSet

In [13]:
sol.args

((1.29957020576342, 1.44984214271483),)

In [14]:
sol.args[0]

(1.29957020576342, 1.44984214271483)

In [15]:
sol.args[0][0]


1.29957020576342

In [16]:
pa = sol.args[0][0]
pc = sol.args[0][1]
print(f' Density of 100% amorphous polymer, pa = {round(pa,2)} g/cm3')
print(f' Density of 100% crystaline polymer, pc = {round(pc,2)} g/cm3')


 Density of 100% amorphous polymer, pa = 1.3 g/cm3
 Density of 100% crystaline polymer, pc = 1.45 g/cm3


## Use SymPy to calculate a numerical result

SymPy expressions can be used to calculate a numerical result if all of the variables in an expression are set to floats or integers. 

Let's solve the following problem with SymPy and calculate a numerical result.

#### Given:

The density of a 100\% amorphous polymer sample $\rho_a$ and the density of a 100% crystaline sample $\rho_c$ of the same polymer are measured. 

$$ \rho_a = 1.30 \ g/cm^3 $$

$$ \rho_c = 1.45 \ g/cm^3 $$

If the density of the same polymer sample $\rho_s$ is measured.

$$ \rho_s =  1.382 \ g/cm^3 $$

### Find:

What is the \% crytallinity of the sample with a measured density $ \rho_s =  1.382 \ g/cm^3 $?

### Solution

We have precise values for $ \rho_a $ and $ \rho_c $ from the previous problem. Let's see what the values of $ \rho_a $ and $ \rho_c $ are. We will use these more precise values that we calculated earlier to solve the problem. 

In [17]:
pa

1.29957020576342

In [18]:
pc

1.44984214271483

Next, we will create three SymPy symbols objects. These three symbols objects will be used to build our expression.

In [19]:
pc, pa, ps = symbols('pc pa ps')

The expression that relates \% crystallinity of a polymer sample to the density of amorphus and crystalline versions of the same polymer is below.

$$ \%crystallinity = \frac{ \rho_c(\rho_s - \rho_a)   }{\rho_s(\rho_c - \rho_a)   } \times 100 \% $$

We will build a SymPy expression that represents the equation above using the symbols objects (variables) we just defined.

In [20]:
expr3 = ( pc*(ps-pa)   ) / (ps*(pc-pa))

Now we can substitute our $ \rho_a $ and $ \rho_c $ from a above. Note the SymPy's ```.subs()``` method does not modify an expression in place. We have to set the modified expression to a new variable so that we can make another substitution. After the substitutions are complete, we can print out the numerical value of the expression. This is accomplished with SymPy's ```.evalf()``` method.

In [21]:
expr3 = expr3.subs(pa, 1.29957020576342)
expr3 = expr3.subs(pc, 1.44984214271483)
expr3 = expr3.subs(ps, 1.382)

print(expr3.evalf())

0.575465123010107


In [22]:
print(f'The percent crystallinity of the sample is {round(expr3*100,2)} percent')

The percent crystallinity of the sample is 57.55 percent
